### PR TITLE
Plumb rclpy.init context to get_default_launch_description

### PR DESCRIPTION
**Pull request by pbaughman**
From https://github.com/ros2/launch/issues/189

This lets you pass a launch context to get_default_launch_description into get_default_launch_description and plumbs it all the way down to the ROSSpecificLaunchStartup launch action.

I couldn't find where rclpy.shutdown is called, so I didn't hook that up anywhere.

I also had to change `rclpy.get_global_executor` to constructing my own SingleThreadedExecutor because the executor needs to be built with the same context as the node.

#### This Incomplete!
This is about 50% of what's needed for issue 189.  The other half is to make ros2 launch actually use this plumbing.  A question for @wjwwood - is it necessary to shut-down the context when we're done, or is it sufficient to just [pass a context here?](https://github.com/ros2/launch/blob/632ef1f3eca7dbc54fbed85ab65a5efc512d83ed/ros2launch/ros2launch/api/api.py#L140-L141)

### CODE REVIEW

File: launch_ros/launch_ros/default_launch_description.py [hunk 1]
```diff
@@ -94,12 +96,12 @@ def _run(self):
 
     def _function(self, context: launch.LaunchContext):
         try:
-            rclpy.init(args=context.argv)
+            rclpy.init(args=context.argv, context=self.__rclpy_context)
```
Comment (wjwwood) [inline launch_ros/launch_ros/default_launch_description.py]: This is only needed if the context is not already initialized. I'd recommend requiring that the context that was passed in is initialized, and also require that shutting it down is the responsibility of the caller.

File: ros2launch/ros2launch/api/api.py [hunk 2]
```diff
@@ -137,8 +139,14 @@ def parse_launch_arguments(launch_arguments: List[Text]) -> List[Tuple[Text, Tex
 def launch_a_python_launch_file(*, python_launch_file_path, launch_file_arguments, debug=False):
     """Launch a given Python launch file (by path) and pass it the given launch file arguments."""
     launch_service = launch.LaunchService(argv=launch_file_arguments, debug=debug)
+    context = rclpy.context.Context()
+    rclpy.init(args=launch_file_arguments, context=context)
```
Comment (pbaughman) [inline ros2launch/ros2launch/api/api.py]: @wjwwood I'm afraid I'm not up on what kind of arguments we're expecting to pass to the context here so I'm not able to test this very well.  It looks like the launch_file_arguments get passed all the way through to become LaunchContext.argv which gets passed to rclpy.init up in ROSSpecificLaunchStartup so this should do the same thing, but I'm afraid I haven't checked

Comment (wjwwood) [reply ros2launch/ros2launch/api/api.py]: I think this is right.

**Comment by wjwwood** [comment]
CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/6329/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/2743/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/5212/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/6179/)

**Comment by wjwwood** [comment]
The test failures look unrelated/flaky:

- the `test_cli` one that is present on Linux and Windows: https://github.com/ros2/build_cop/issues/166
- the `gtest_waitable` one that is present on macOS is listed here: https://github.com/ros2/build_cop/issues/168

**Pull request merged**